src/memory_region.rs

```diff
@@ -155,20 +147,165 @@ pub enum AccessType {
+            }
+        }
+
+        let mut result = Self {
+            regions: vec![MemoryRegion::default(); regions.len()].into_boxed_slice(),
+            region_addresses: vec![0; regions.len()].into_boxed_slice(),
+            cache: UnsafeCell::new(MappingCache::new()),
+            config,
+        };
+        result.construct_eytzinger_order(&regions, 0, 0);
+        Ok(result)
+    }
+
+    /// Given a list of regions translate from virtual machine to host address
+    #[allow(clippy::integer_arithmetic)]
+    pub fn map<E: UserDefinedError>(
+        &self,
+        access_type: AccessType,
+        vm_addr: u64,
+        len: u64,
+    ) -> Result<u64, EbpfError<E>> {
+        // Safety:
+        // &mut references to the mapping cache are only created internally here
+        // and in replace_region(). The methods never invoke each other and
+        // UnalignedMemoryMapping is !Sync, so the cache reference below is
+        // guaranteed to be unique.
+        let cache = unsafe { &mut *self.cache.get() };
+        let (cache_miss, index) = if let Some(region) = cache.find(vm_addr) {
+            (false, region)
+        } else {
+            let mut index = 1;
+            while index <= self.region_addresses.len() {
+                // Safety:
+                // we start the search at index=1 and in the loop condition check
+                // for index <= len, so bound checks can be avoided
+                index = (index << 1)
+                    + unsafe { *self.region_addresses.get_unchecked(index - 1) <= vm_addr }
+                        as usize;
+            }
+            index >>= index.trailing_zeros() + 1;
+            if index == 0 {
+                return generate_access_violation(self.config, access_type, vm_addr, len);
+            }
+            (true, index)
+        };
+
+        // Safety:
+        // we check for index==0 above, and by construction if we get here index
+        // must be contained in region
+        let region = unsafe { self.regions.get_unchecked(index - 1) };
+        if access_type == AccessType::Load || region.is_writable {
+            if let Ok(host_addr) = region.vm_to_host::<E>(vm_addr, len as u64) {
+                if cache_miss {
+                    cache.insert(
+                        region.vm_addr..region.vm_addr.saturating_add(region.len),
+                        index,
+                    );
+                }
+                return Ok(host_addr);
+            }
+        }
+
+        generate_access_violation(self.config, access_type, vm_addr, len)
+    }
+
+    /// Returns the `MemoryRegion`s in this mapping
+    pub fn get_regions(&self) -> &[MemoryRegion] {
+        &self.regions
+    }
+
+    /// Replaces the `MemoryRegion` at the given index
+    pub fn replace_region<E: UserDefinedError>(
+        &mut self,
+        index: usize,
+        region: MemoryRegion,
+    ) -> Result<(), EbpfError<E>> {
+        if index >= self.regions.len() {
```

Comment:
We have to make sure that the `vm_addr` stays the same so that the order of the memory region stays stable.